# Pandas3 - Data Preprocessing and Calculation

In [1]:
#import pacakge
import pandas as pd

## (1) Oil price

**(1.1) Given the following dataframe, calcualte the number of missing values.**

In [2]:
month = ['Jan 2018','Feb 2018','Mar 2018','Apr 2018','May 2018','Jun 2018','Jul 2018','Aug 2018','Sep 2018','Oct 2018','Nov 2018','Dec 2018','Jan 2019','Feb 2019','Mar 2019','Apr 2019','May 2019','Jun 2019','Jul 2019','Aug 2019','Sep 2019','Oct 2019','Nov 2019','Dec 2019']
price = [68.5,63.3,63.7,69.1,74.9,73.4,None,71.7,78.1,79.3,64.8,57.6,None,63.9,66,71.5,70.9,61.9,None,59.4,None,58.5,63.3,64.5]
oil_df = pd.DataFrame({"month":month,"price":price})

In [3]:
# put your code here
oil_df.isnull().sum()

month    0
price    4
dtype: int64

**(1.2) Use the dataframe in (1.1). Fill in missing values with the price of the previous month.**

In [4]:
# put your code here
oil_df.fillna(method = 'ffill')

month  price
0   Jan 2018   68.5
1   Feb 2018   63.3
2   Mar 2018   63.7
3   Apr 2018   69.1
4   May 2018   74.9
5   Jun 2018   73.4
6   Jul 2018   73.4
7   Aug 2018   71.7
8   Sep 2018   78.1
9   Oct 2018   79.3
10  Nov 2018   64.8
11  Dec 2018   57.6
12  Jan 2019   57.6
13  Feb 2019   63.9
14  Mar 2019   66.0
15  Apr 2019   71.5
16  May 2019   70.9
17  Jun 2019   61.9
18  Jul 2019   61.9
19  Aug 2019   59.4
20  Sep 2019   59.4
21  Oct 2019   58.5
22  Nov 2019   63.3
23  Dec 2019   64.5

**(1.3) Use the dataframe in (1.1). Fill in missing values with the average price from 2018 to 2019.**

In [5]:
# put your code here
oil_df.fillna(oil_df.mean())

month   price
0   Jan 2018  68.500
1   Feb 2018  63.300
2   Mar 2018  63.700
3   Apr 2018  69.100
4   May 2018  74.900
5   Jun 2018  73.400
6   Jul 2018  67.215
7   Aug 2018  71.700
8   Sep 2018  78.100
9   Oct 2018  79.300
10  Nov 2018  64.800
11  Dec 2018  57.600
12  Jan 2019  67.215
13  Feb 2019  63.900
14  Mar 2019  66.000
15  Apr 2019  71.500
16  May 2019  70.900
17  Jun 2019  61.900
18  Jul 2019  67.215
19  Aug 2019  59.400
20  Sep 2019  67.215
21  Oct 2019  58.500
22  Nov 2019  63.300
23  Dec 2019  64.500

## (2) Melbourne housing data

**(2.1) Import the dataset <code>melbourne.csv</code>. Show the first five rows.**

In [6]:
# put your code here
melbourne_df = pd.read_csv("../dataset/melbourne.csv")
melbourne_df.head(5)

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
3  4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0           NaN   
4  4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
3        NaN        Yarra  -37.7969    144.9969  Northern Metropolitan   
4     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

**(2.2) Select the data for the following suburbs. Store the result as another data frame.** <br>
<code>Reservoir, Richmond, Brunswick, Bentleigh East, Coburg, Preston</code>.

In [7]:
# put your code here
melbourne_sub = melbourne_df[melbourne_df.Suburb.isin(['Reservoir', 'Richmond', 'Brunswick', 'Bentleigh East', 'Coburg', 'Preston'])]
melbourne_sub.head(5)

Suburb          Address  Rooms Type      Price Method  \
787  Bentleigh East     18 Mervin St      2    h  1410000.0      S   
788  Bentleigh East  1/69 Purtell St      3    t   760000.0     PI   
789  Bentleigh East   30 Veronica St      2    h   825000.0      S   
790  Bentleigh East     2a Bessie St      2    u   746000.0      S   
791  Bentleigh East    14a Daphne St      1    u   580000.0      S   

           SellerG       Date  Distance  Postcode  ...  Bathroom  Car  \
787  hockingstuart  3/09/2016      13.9    3165.0  ...       1.0  2.0   
788  hockingstuart  3/09/2016      13.9    3165.0  ...       1.0  2.0   
789         Buxton  3/09/2016      13.9    3165.0  ...       2.0  1.0   
790  hockingstuart  3/12/2016      13.9    3165.0  ...       1.0  2.0   
791       Woodards  3/12/2016      13.9    3165.0  ...       1.0  1.0   

     Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
787     768.0           NaN        NaN    Glen Eira  -37.9305    145.0547   
788     287.0           NaN        NaN    Glen Eira  -37.9261    145.0566   
789     176.0         125.0     2008.0    Glen Eira  -37.9237    145.0763   
790     281.0          95.0     2002.0    Glen Eira  -37.9255    145.0796   
791     105.0           NaN        NaN    Glen Eira  -37.9181    145.0586   

                Regionname Propertycount  
787  Southern Metropolitan       10969.0  
788  Southern Metropolitan       10969.0  
789  Southern Metropolitan       10969.0  
790  Southern Metropolitan       10969.0  
791  Southern Metropolitan       10969.0  

[5 rows x 21 columns]

**(2.3) Calculate the average price in each suburb. Which suburb has the highest average house price?**<br>
Add your answers to the markdown cell.

In [8]:
# [solution-1]

#step1: Get groupby object
gb_suburb = melbourne_sub.groupby("Suburb")

#step2: Get statistics (average) of all numerical columns (return a dataframe)
average_df = gb_suburb.mean()

#step3: Sort the dataframe
average_df.sort_values(by = "Price", ascending = False)

Rooms         Price   Distance  Postcode  Bedroom2  \
Suburb                                                                  
Bentleigh East  3.152610  1.085592e+06  13.881928    3165.0  3.100402   
Richmond        2.384615  1.083564e+06   2.560769    3121.0  2.373077   
Brunswick       2.648649  1.013171e+06   5.200000    3056.0  2.621622   
Coburg          2.973684  9.224936e+05   7.574211    3058.0  2.905263   
Preston         2.920502  9.028008e+05   8.706276    3072.0  2.899582   
Reservoir       2.810585  6.900089e+05  11.420613    3073.0  2.777159   

                Bathroom       Car    Landsize  BuildingArea    YearBuilt  \
Suburb                                                                      
Bentleigh East  1.598394  1.867470  479.578313    165.049675  1975.770370   
Richmond        1.392308  1.054054  564.369231     98.117568  1959.411765   
Brunswick       1.301802  1.162896  316.954955    124.170956  1942.081081   
Coburg          1.394737  1.478947  379.821053    129.312696  1952.155738   
Preston         1.330544  1.690377  459.581590    131.169910  1954.774194   
Reservoir       1.284123  1.849582  584.278552    124.555345  1975.073892   

                Lattitude  Longtitude  Propertycount  
Suburb                                                
Bentleigh East -37.920656  145.067790        10969.0  
Richmond       -37.820303  145.001924        14949.0  
Brunswick      -37.766591  144.961853        11918.0  
Coburg         -37.745084  144.962669        11204.0  
Preston        -37.740418  145.004753        14577.0  
Reservoir      -37.713984  145.009586        21650.0

In [9]:
# [solution-2] Select columns before groupby (return a dataframe)
melbourne_sub.loc[:,["Suburb","Price"]].groupby("Suburb").mean().sort_values(by = 'Price', ascending = False)

Price
Suburb                      
Bentleigh East  1.085592e+06
Richmond        1.083564e+06
Brunswick       1.013171e+06
Coburg          9.224936e+05
Preston         9.028008e+05
Reservoir       6.900089e+05

In [10]:
# [solution-3] Select column after groupby (return a series)
melbourne_sub.groupby("Suburb").Price.mean().sort_values(ascending = False)

Suburb
Bentleigh East    1.085592e+06
Richmond          1.083564e+06
Brunswick         1.013171e+06
Coburg            9.224936e+05
Preston           9.028008e+05
Reservoir         6.900089e+05
Name: Price, dtype: float64

**Write you answers here:**<br>
The average house price in Bentleigh East is the highest.

**(2.4) What is the largest </code>land size</code> in each suburb?**

In [11]:
# put your code here
melbourne_sub.loc[:,["Suburb","Landsize"]].groupby("Suburb").max()

Landsize
Suburb                  
Bentleigh East    1080.0
Brunswick         4676.0
Coburg            3747.0
Preston           3037.0
Reservoir        41400.0
Richmond         14294.0

**(2.5) What is the highest price in each region (use column <code>Regionname</code>).**

In [12]:
# put your code here
melbourne_sub.loc[:,[ "Regionname","Price"]].groupby("Regionname").max()

Price
Regionname                      
Northern Metropolitan  3335000.0
Southern Metropolitan  1962000.0

## (3) US National Parks and their animal and plant species
The National Park Service publishes a database of animal and plant species identified in individual national parks and verified by evidence — observations, vouchers, or reports that document the presence of a species in a park. All park species records are available to the public on the National Park Species portal; exceptions are made for sensitive, threatened, or endangered species when widespread distribution of information could pose a risk to the species in the park.

National Park species lists provide information on the presence and status of species in US national parks. These species lists are works in progress and the absence of a species from a list does not necessarily mean the species is absent from a park. The time and effort spent on species inventories varies from park to park, which may result in data gaps. Species taxonomy changes over time and reflects regional variations or preferences; therefore, records may be listed under a different species name.

Each park species record includes a species ID, park name, taxonomic information, scientific name, one or more common names, record status, occurrence (verification of species presence in park), nativeness (species native or foreign to park), abundance (presence and visibility of species in park), seasonality (season and nature of presence in park), and conservation status (species classification according to US Fish & Wildlife Service). Taxonomic classes have been translated from Latin to English for species categorization; order, family, and scientific name (genus, species, subspecies) are in Latin.

**(3.1) Import the dataset <code>species.csv</code>. Show the summary of this dataset including column names and their data types.**

In [13]:
# put your code here
species_raw_df = pd.read_csv("../dataset/species.csv")
species_raw_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119248 entries, 0 to 119247
Data columns (total 14 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   Species ID           119248 non-null  object
 1   Park Name            119248 non-null  object
 2   Category             119248 non-null  object
 3   Order                117776 non-null  object
 4   Family               117736 non-null  object
 5   Scientific Name      119248 non-null  object
 6   Common Names         119248 non-null  object
 7   Record Status        119248 non-null  object
 8   Occurrence           99106 non-null   object
 9   Nativeness           94203 non-null   object
 10  Abundance            76306 non-null   object
 11  Seasonality          20157 non-null   object
 12  Conservation Status  4718 non-null    object
 13  Unnamed: 13          5 non-null       object
dtypes: object(14)
memory usage: 12.7+ MB


C:\BI-Apps\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3165: DtypeWarning: Columns (13) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


**(3.2) Create a subset named <code>species_df</code> with the following columns: <code>'Species ID', 'Park Name', 'Common Names', 'Family', 'Occurrence', 'Nativeness', 'Abundance'</code>.**

In [14]:
# put your code here
species_df = species_raw_df.loc[:,['Species ID','Park Name', 'Common Names', 'Family', 'Occurrence', 'Nativeness','Abundance']]
species_df.head(10)

Species ID             Park Name  \
0  ACAD-1000  Acadia National Park   
1  ACAD-1001  Acadia National Park   
2  ACAD-1002  Acadia National Park   
3  ACAD-1003  Acadia National Park   
4  ACAD-1004  Acadia National Park   
5  ACAD-1005  Acadia National Park   
6  ACAD-1006  Acadia National Park   
7  ACAD-1007  Acadia National Park   
8  ACAD-1008  Acadia National Park   
9  ACAD-1009  Acadia National Park   

                                        Common Names      Family  \
0                                              Moose    Cervidae   
1  Northern White-Tailed Deer, Virginia Deer, Whi...    Cervidae   
2                             Coyote, Eastern Coyote     Canidae   
3        Eastern Timber Wolf, Gray Wolf, Timber Wolf     Canidae   
4  Black Fox, Cross Fox, Eastern Red Fox, Fox, Re...     Canidae   
5                                        Canada Lynx     Felidae   
6               Bay Lynx, Bobcat, Red Lynx, Wild Cat     Felidae   
7              Eastern Skunk, Polecat, Striped Skunk  Mephitidae   
8                                 Otter, River Otter  Mustelidae   
9                            Blackcat, Fisher, Pekan  Mustelidae   

                        Occurrence  Nativeness Abundance  
0                          Present      Native      Rare  
1                          Present      Native  Abundant  
2                          Present  Not Native    Common  
3                    Not Confirmed      Native       NaN  
4                          Present     Unknown    Common  
5                    Not Confirmed      Native       NaN  
6  Not Present (Historical Report)  Not Native       NaN  
7  Not Present (Historical Report)      Native       NaN  
8                          Present      Native    Common  
9                          Present      Native      Rare

**(3.3) Replace missing values with "Unknown".**

In [15]:
# put your code here
species_df.fillna("Unknown", inplace = True)

**(3.4) Shows the number of species in each "abundance" category.**<br>
Example:<br>
Uncommon:   18779<br>
Common:     12913<br>
...

In [16]:
# [solution-1] Select the categorical column and calculate the frequency of each unique value (sort in descending order)
species_df.Abundance.value_counts()

Unknown       71061
Uncommon      18779
Common        12913
Rare          10768
Occasional     3312
Abundant       2365
Native           40
Not Native       10
Name: Abundance, dtype: int64

In [17]:
# [solution-2] Group the data by the categorical column and count the number of rows (species) in each group 
species_df.groupby("Abundance").size()

Abundance
Abundant       2365
Common        12913
Native           40
Not Native       10
Occasional     3312
Rare          10768
Uncommon      18779
Unknown       71061
dtype: int64

**(3.5) Grouping data based on column <code>Park Name</code>.**

In [18]:
# put your code here
gb_park = species_df.groupby("Park Name")

**(3.6) Use the groupby object obtained in (1.5). Calculate the number of species in each park and show the top five parks with the largest number of species.**

In [19]:
# put your code here
gb_park.size().sort_values(ascending = False).head(5)

Park Name
Great Smoky Mountains National Park    6623
Redwood National Park                  6310
Shenandoah National Park               4655
Death Valley National Park             4439
Yellowstone National Park              3966
dtype: int64

**(3.7) Use the groupby object obtained in (3.5). Calculate the number of species in each park and store the result in a dataframe named <code>species_park_df</code>. The dataframe should contains two columns:<code>Park Name</code> and <code>number</code>**<br>
For example:

||Park Name|Number_of_species|
|:-:|:-:|:-:|
|**0**|Acadia National Park|1709|
|**1**|Arches National Park|1048|
|**2**|Badlands National Park|1389|
|**3**|...|...|
|**4**|...|...|

In [20]:
# put your code here
species_park_df = gb_park.size().to_frame(name = "Number_of_species").reset_index()
species_park_df.head(10)

Park Name  Number_of_species
0                        Acadia National Park               1709
1                        Arches National Park               1048
2                      Badlands National Park               1389
3                      Big Bend National Park               2269
4                      Biscayne National Park               1726
5  Black Canyon of the Gunnison National Park               1106
6                  Bryce Canyon National Park               1286
7                   Canyonlands National Park               1223
8                  Capitol Reef National Park               1566
9              Carlsbad Caverns National Park               1536

**(3.8) Import the dataset <code>parks.csv</code> as a dataframe named <code>park_df</code>. Show the first five rows.**

In [21]:
# put your code here
park_df = pd.read_csv("../dataset/parks.csv")
park_df.head(5)

Park Code               Park Name State   Acres  Latitude  Longitude
0      ACAD    Acadia National Park    ME   47390     44.35     -68.21
1      ARCH    Arches National Park    UT   76519     38.68    -109.57
2      BADL  Badlands National Park    SD  242756     43.75    -102.50
3      BIBE  Big Bend National Park    TX  801163     29.25    -103.25
4      BISC  Biscayne National Park    FL  172924     25.65     -80.08

**(3.9) Use <code>species_park_df</code> obtained in (3.7). Add the acres as a new column.**<br>
For example:

||Park Name|Number_of_species|Acres|
|:-:|:-:|:-:|:-:|
|**0**|Acadia National Park|1709|47390|
|**1**|Arches National Park|1048|76519|
|**2**|Badlands National Park|1389|242756|
|**3**|...|...|...|
|**4**|...|...|...|

In [22]:
# put your code here
species_park_df = pd.merge(species_park_df, park_df.loc[:,["Park Name","Acres"]], how = 'left', on = "Park Name")
species_park_df.head(10)

Park Name  Number_of_species   Acres
0                        Acadia National Park               1709   47390
1                        Arches National Park               1048   76519
2                      Badlands National Park               1389  242756
3                      Big Bend National Park               2269  801163
4                      Biscayne National Park               1726  172924
5  Black Canyon of the Gunnison National Park               1106   32950
6                  Bryce Canyon National Park               1286   35835
7                   Canyonlands National Park               1223  337598
8                  Capitol Reef National Park               1566  241904
9              Carlsbad Caverns National Park               1536   46766

**(3.10) Use the result obtained in (3.9). Calculate the <code>number of species per acre</code> in each park and list the five national parks with the highest number of species per acre.**

In [23]:
# put your code here
#step1: Add a new column
species_park_df["Num_species_acre"] = species_park_df.Number_of_species/species_park_df.Acres
species_park_df.head(10)

Park Name  Number_of_species   Acres  \
0                        Acadia National Park               1709   47390   
1                        Arches National Park               1048   76519   
2                      Badlands National Park               1389  242756   
3                      Big Bend National Park               2269  801163   
4                      Biscayne National Park               1726  172924   
5  Black Canyon of the Gunnison National Park               1106   32950   
6                  Bryce Canyon National Park               1286   35835   
7                   Canyonlands National Park               1223  337598   
8                  Capitol Reef National Park               1566  241904   
9              Carlsbad Caverns National Park               1536   46766   

   Num_species_acre  
0          0.036062  
1          0.013696  
2          0.005722  
3          0.002832  
4          0.009981  
5          0.033566  
6          0.035887  
7          0.003623  
8          0.006474  
9          0.032844

In [24]:
#step2: Sort the dataframe
species_park_df.sort_values(by = "Num_species_acre", ascending = False).head(5)

Park Name  Number_of_species   Acres  Num_species_acre
29      Hot Springs National Park               1950    5550          0.351351
27        Haleakala National Park               2580   29094          0.088678
11         Congaree National Park               2321   26546          0.087433
13  Cuyahoga Valley National Park               1941   32950          0.058907
44          Redwood National Park               6310  112512          0.056083